rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,237 @@
+  "log_type": "BIND_DNS",
+  "entries": [...]
+}
+```
+
+`log_type` has to be one of the values returned by the [logtypes endpoint](https://cloud.google.com/chronicle/docs/reference/ingestion-api#logtypes).
+
+
+If `event_type == "unstructured"` we will require another configuration field -
+`log_type` to be set. This field will be templatable. The field can be removed
+from the final event via `encoding.exclude_fields`.
+
+The encoding of the message (text or json) can be set via an `encoding` field.
+
+##### Schema
+
+A UDM message has the following sections. (To keep this RFC succinct, this is
+not a comprehensive list, the full list can be found
+[here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#udm_event_data_model).)
+
+- *Metadata*
+  Contains metadata about the event. The fields available for metadata can be
+  found [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#metadata).
+  Mandatory fields are `event_type` and `event_timestamp`.
+
+  The value of `event_type` has repercussions for which fields are mandatory
+  in the rest of the event. See [here](https://cloud.google.com/chronicle/docs/unified-data-model/udm-usage#required_and_optional_fields_based_on_event_type)
+  for the list. For example, if `event_type == "FILE_COPY"` then `src.file`
+  becomes mandatory. **Currently it is not possible to validate a field based
+  on the value of another field within Vector schema, so this will need to
+  be left for future improvement.
+
+- *Principal*
+  The principal is the entity that originates the event. This field is mandatory
+  and can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Src*
+  The source entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Target*
+  The target entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+### Implementation
+
+Both event types require different batching strategies.
+
+#### UDM
+
+UDM can use a simple `BatchedHttpSink` with a `JsonArrayBuffer` to perform the
+batching.
+
+#### Unstructured log events
+
+Since log events are batched by `log_type` we will need to partition the
+batches on the value of the `log_type` field using a `PartitionHttpSink`.
+
+
+#### Error handling
+
+According to the Chronicle docs:
+
+> Clients should retry on 500 and 503 errors with exponential backoff. The
+> minimum delay should be 1s unless it is documented otherwise. For 429
+> errors, the client may retry with minimum 30s delay. For all other errors,
+> retry may not be applicable.
+
+I believe the current GCP sinks implement this behaviour so we can reuse the
+logic from these sinks.
```

Comment:
@vladimir-dd Are you able to confirm if this is the case?

Reply:
From what I can see I'm not sure we would be waiting for minimum 30 sec for 429 by default, so we might need to tune retry settings a bit. @tobz should know more about it.

Reply:
@bruceg has also been working in the GCP source/sinks recently.